lib/Analysis/Memory/DIDependenceAnalysis.cpp

```diff
@@ -590,6 +590,67 @@ trait::Reduction::Kind getReductionKind(
   return trait::DIReduction::RK_NoReduction;
 }
 
+bool handleLoopEmptyBindings(
+    const Loop *L, const DIMemoryTrait &DITraitItr,
+    DIMemoryTraitRegionPool &Pool,
+    const SpanningTreeRelation<const tsar::DIAliasTree *> &DIAliasSTR) {
+  auto *Mem = DITraitItr.getMemory();
+  if (Mem->emptyBinding())
+    return true;
+  for (auto &Loc : *Mem) {
+    if (!Loc.pointsToAliveValue())
+      continue;
+    for (auto *User : Loc->users())
+      if (auto *I = dyn_cast<Instruction>(User))
+        if (L->contains(I)) {
+          return false;
+        }
+    if (auto *CE = dyn_cast<ConstantExpr>(Loc)) {
+      SmallVector<ConstantExpr *, 4> WorkList{CE};
+      do {
+        auto *Expr = WorkList.pop_back_val();
+        for (auto &ExprU : Expr->uses()) {
+          auto ExprUse = ExprU.getUser();
+          if (auto ExprUseInst = dyn_cast<Instruction>(ExprUse))
+            if (L->contains(ExprUseInst))
+              return false;
+            else if (auto ExprUseExpr = dyn_cast<ConstantExpr>(ExprUse))
+              WorkList.push_back(ExprUseExpr);
+        }
+      } while (!WorkList.empty());
+    }
```

Comment:
Здесь что-то странное было со вложенностью `if` в цикл и в друг друга.

```suggestion
    for (auto *User : Loc->users()) {
      if (auto *I = dyn_cast<Instruction>(User)) {
        if (L->contains(I))
          return false;
      } else if (auto *CE = dyn_cast<ConstantExpr>(User)) {
        SmallVector<ConstantExpr *, 4> WorkList{CE};
        do {
          auto *Expr = WorkList.pop_back_val();
          for (auto &ExprU : Expr->uses()) {
            auto ExprUse = ExprU.getUser();
            if (auto ExprUseInst = dyn_cast<Instruction>(ExprUse)) {
              if (L->contains(ExprUseInst))
                return false;
            } else if (auto ExprUseExpr = dyn_cast<ConstantExpr>(ExprUse)) {
              WorkList.push_back(ExprUseExpr);
            }
          }
        } while (!WorkList.empty());
      }
    }
```